lsコマンドのl オプション完成

Yatsu0720 opened the pull request:
lsコマンドのlオプションが完成しましたので、確認お願いいたします｡

Yatsu0720 committed:
各要素の配列を作成

Yatsu0720 committed:
データ格納方法を配列からハッシュに変更

Yatsu0720 committed:
各要素の最大文字数を取得するようにした

Yatsu0720 committed:
lコマンド（引数で指定は未）一旦完成

Yatsu0720 committed:
オプションなしの場合にも対応

Yatsu0720 committed:
lコマンドの引数にも対応し、完成

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

Yatsu0720 committed:
# On branch ll
# Changes to be committed:
#   (use "git restore --staged <file>..." to unstage)
# 	modified:   05.ls/ls.rb
#
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
# 	05.ls/dummyfile
# 	05.ls/gemfile
# 	05.ls/gemfile:w
# 	"05.ls/ls\343\203\225\343\203\255\343\203\274\343\203\201\343\203\243\343\203\274\343\203\210.drawio"
# 	05.ls/test/

Yatsu0720 commented on 05.ls/ls.rb:
```diff
@@ -5,63 +5,119 @@
-    file_list.sort do |a,b|
-      
-    end.reverse
+def aquire_directory
```
「ディレクトリを取得する」という意図で命名しようとしたのですが、`get`はあまり使わないほうが良いとのことなので、迷いましたがあまり聞きなれなかったaquireをWeb辞書で調べて選択しました。

Yatsu0720 commented:
いただいた指摘事項を踏まえ修正しましたのでご確認お願いいたします。

JunichiIto reviewed (commented):
今回も動画でレビューしたので別途リンクを送ります〜。

Yatsu0720 committed:
8進数への変換をDRYに

Yatsu0720 committed:
opt.parseの実装

Yatsu0720 committed:
marginを単に数字を書き入れるように変更

Yatsu0720 committed:
メソッドを一回ずつ読むように修正

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -9,19 +9,22 @@
-  if options['l']
-    puts("total #{total_block_size}")
-    print_file_list(file_stats)
+  l_command = false
```
l_commandじゃなくて、l_optionと呼んであげる方が正しいかも

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -35,89 +38,74 @@ def output_no_option
-    ARGV[0]
-  end
+def fetch_files
+  opts = OptionParser.new
```
mainメソッドの中でOptionParserを使ってるので、OptionParserを使うのはそこだけにしたい（起動時引数の処理は一箇所にまとめたい）です。

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -35,89 +38,74 @@ def output_no_option
+  puts("total #{total_block_size}")
+  file_stats.each do |file_stat|
+    print file_stat[:permission]
+    print "#{file_stat[:hard_link].rjust(hard_link_padding)} "
```
calculate_max_lengthをこのメソッド内で呼んで、

```suggestion
    print "#{file_stat[:hard_link].rjust(max_lengths[:hard_link] + 1)} "
```

とした方がシンプルかなと思いました。

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -35,89 +38,74 @@ def output_no_option
-    print file_list[:time]
-    puts file_list[:file_name]
+  total_block_size = file_stats.sum { |block_size| block_size[:blocks_size] }
+  puts("total #{total_block_size}")
```
```suggestion
  puts "total #{total_block_size}"
```

他のputsやprintには()が付いてないので。

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -9,19 +9,22 @@
+    print_files(file_stats)
   else
-    output_no_option
+    output_no_option(file_stats)
```
print_filesのprintとoutput_no_optionのoutputは同じ処理を指していると思いますが、呼び方が違うのが気になります〜。
命名に一貫性を持たせましょう。

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -35,89 +38,74 @@ def output_no_option
+  opts.on('-l')
+  opts.parse!(ARGV) || '.'
+  ls_dir ||= ARGV[0] || '.'
+  Dir.chdir(ls_dir)
```
`Dir.chdir do`の形にして、ディレクトリが変わる範囲を限定した方がいいと思います。
でないと、`fetch_files`メソッドが予期しない副作用を持ってしまうので。

https://docs.ruby-lang.org/ja/latest/method/Dir/s/chdir.html

JunichiIto commented on 05.ls/lsフローチャート.drawio:
```diff
@@ -0,0 +1 @@
+<mxfile host="app.diagrams.net" modified="2022-06-08T10:23:36.685Z" agent="5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.162 Safari/537.36" etag="IGA8RJgbbLQn8z95Zbnz" version="19.0.2" type="device"><diagram id="8puRApz0o0APBaA4m-Bt" name="Page-1">7VhLc5swEP41OsYDCDAcwY9mptNOZnJoe6RGNjTYcmT51V/fFVrxsmnTunGmaU5erb7dlfbxRYHQ0fLwTiTr7ANPWUEcKz0QOiaO ...
```
このファイルやdummyfileなど、動作確認用のファイルがメインプログラムと同じディレクトリにあると、ファイルの重要性がわからない（このファイルもプログラムの動作に必要なのかな？と勘違いされる）ので、「明らかに動作確認用だろう」とわかるディレクトリに置いてある方がいいです。

たとえば次のようにすれば、カレントディレクトリのファイル一覧を表示できるのではないでしょうか？

```
cd test
ruby ../ls.rb
```

JunichiIto reviewed (commented):
コメントしましたー。

Yatsu0720 committed:
各種指摘事項の修正

Yatsu0720 committed:
不要なライブラリを削除

Yatsu0720 commented:
コメントありがとうございました。
修正いたしましたので確認お願いいたします！

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -83,29 +74,30 @@ def convert_file_stat_chars(file_stat)
-  opts = OptionParser.new
-  opts.on('-l')
-  opts.parse!(ARGV) || '.'
   ls_dir ||= ARGV[0] || '.'
```
ARGVを登場させるのはmainメソッドだけにして、この行にあるARGVはなくしてみましょう

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -83,29 +74,30 @@ def convert_file_stat_chars(file_stat)
-    file_stats[:time] = file_stat.mtime.strftime(' %b %e %R ')
-    file_stats[:file_name] = File.basename(file)
-    file_stats
+    Dir.chdir(ls_dir) do
```
`file`がディレクトリ情報を含んでいれば（`"foo/bar.txt"`のようになっていれば）、chdirしなくて済みそうだな〜と思いました。

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -38,37 +38,28 @@ def output_no_option(file_stats)
-    max_lengths[:file_size] = [max_lengths[:file_size].to_i.size, hash[:file_size].to_i.size].max
+    max_lengths[:hard_link] = [max_lengths[:hard_link].to_i, hash[:hard_link].to_i].max.to_s.size
+    max_lengths[:user] = [max_lengths[:user].to_s, hash[:user]].max.size
+    max_lengths[:group] = [max_lengths[:group].to_s, hash[:group]].max.size
```
60と61行目、ロジック的におかしそうな気がします。

Yatsu0720 replied:
`:hardlink`や`file_size`の部分もちょっと処理がおかしいと感じたので、あわせて直しました！

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -83,29 +74,30 @@ def convert_file_stat_chars(file_stat)
+      file_stats[:hard_link] = file_stat.nlink.to_s
+      file_stats[:user] = Etc.getpwuid(file_stat.uid).name
+      file_stats[:group] = Etc.getgrgid(file_stat.gid).name
+      file_stats[:file_size] = file_stat.size.to_s
```
hard_linkやfile_sizeは数値なので、

```suggestion
      file_stats[:file_size] = file_stat.size
```

のように数値のまま保持した方が再利用性が高くなります。
（`to_s`したものを62行目で再び`to_i`で数値に戻したりするのは無駄な型変換です）

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -83,29 +74,30 @@ def convert_file_stat_chars(file_stat)
+      file_stats[:user] = Etc.getpwuid(file_stat.uid).name
+      file_stats[:group] = Etc.getgrgid(file_stat.gid).name
+      file_stats[:file_size] = file_stat.size.to_s
+      file_stats[:time] = file_stat.mtime.strftime(' %b %e %R ')
```
timeもTime型で保持し続けて、51行目で整形する方がやはりデータの再利用性の面でベターです。

JunichiIto reviewed (commented):
コメントしました！

Yatsu0720 committed:
追加の指摘事項を修正

Yatsu0720 committed:
rubocop対応

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -14,8 +14,9 @@ def main
   opts = OptionParser.new
   opts.on('-l') { l_option = true }
   opts.parse!(ARGV) || '.'
+  ls_dir ||= ARGV[0] || '.'
```
`ls_dir`はここで初めて宣言された変数なので`=`で良いはずです

Yatsu0720 replied:
確かにそうでした！　修正しました。

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -73,30 +74,28 @@ def convert_file_stat_chars(file_stat)
   Dir.chdir(ls_dir) do
-    files = Dir.glob('*').sort
-    return files, ls_dir
+    Dir.glob('*').sort.map do |file|
```
globに渡す引数でディレクトリを指定した方がシンプルになりそうです。

https://docs.ruby-lang.org/ja/latest/method/Dir/s/=5b=5d.html

Yatsu0720 replied:
リファレンスは読んでいたのですが、globの使い方を理解しきれていませんでした。
何度か試してみて、書き方がわかりました！

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -73,30 +74,28 @@ def convert_file_stat_chars(file_stat)
-    files = Dir.glob('*').sort
-    return files, ls_dir
+    Dir.glob('*').sort.map do |file|
+      "#{ls_dir}/#{file}"
```
`ruby ls.rb test/` みたいに引数が渡されると、`test//foo.txt`みたいなパスができあがったりしませんか？

Yatsu0720 replied:
確かにそうですね……。スラッシュ付きの場合の確認をしておりませんでした。
今回の修正で、スラッシュ付きでも問題なく動くこと確認しました！

JunichiIto reviewed (commented):
コメントしました！

Yatsu0720 committed:
globの使い方を見直し

Yatsu0720 committed:
debug用コードの削除

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -92,11 +93,7 @@ def make_file_stats(ls_dir)
-      "#{ls_dir}/#{file}"
-    end
-  end
+  Dir.glob("#{ls_dir}/*").sort
```
今回はこれでOKですが、File.joinを使うとより安全に（？）パスを連結することができますー。

https://docs.ruby-lang.org/ja/latest/method/File/s/join.html

JunichiIto commented on 05.ls/ls.rb:
```diff
@@ -5,63 +5,94 @@
+end
 
+def make_file_stats(ls_dir)
+  files = fetch_files(ls_dir)
```
```suggestion
  files = Dir.glob("#{ls_dir}/*").sort
```

メソッド化するまでもないかも？

JunichiIto reviewed (approved):
数点コメントしましたがこれでOKとします！
必要に応じてls5で修正してください〜。
